File: temperature-front/src/app/services/temperature.service.ts
import { HttpClient, HttpResponse } from '@angular/common/http';
import { Injectable } from '@angular/core';
import { Observable } from 'rxjs';
import { environment } from 'src/environments/environment';
import { SensorData } from '../model/SensorData.model';
import { Temperature } from '../model/Temperature.model';

@Injectable({
  providedIn: 'root'
})
export class TemperatureService {
  private url = environment.apiUrl + "temperatures";

  constructor(private httpClient: HttpClient) { }

  getRecentTemperatures(): Observable<Temperature[]> {
    return this.httpClient.get<Temperature[]>(this.url);
  }

  getDeviceTemperatures(deviceId: number): Observable<Temperature[]> {
    return this.httpClient.get<Temperature[]>(this.url + `/devices/${deviceId}`);
  }

  reportTemperatures(sensorsData: SensorData): Observable<HttpResponse<any>> {
    return this.httpClient.post<any>(this.url, sensorsData, {observe: "response"});
  }
}

File: temperature-api/src/main/java/io/github/mahmoudalikhalil/temperatureapi/service/TemperatureService.java
package io.github.mahmoudalikhalil.temperatureapi.service;

import io.github.mahmoudalikhalil.temperatureapi.dto.SensorDTO;
import io.github.mahmoudalikhalil.temperatureapi.dto.TemperatureDTO;

import java.util.List;

public interface TemperatureService {
    void reportTemperatures(SensorDTO sensorData);

    List<TemperatureDTO> getRecentTemperatures();

    List<TemperatureDTO> getDeviceTemperatures(Long deviceId);
}
